### ISSUE TITLE
Huge overhead for each test file loaded.

### ISSUE THREAD

**Issue by AlexMcConnell** [dart-lang/test#1567]
I first noticed this and commented on an issue in the flutter repo before tracing it back up to here (I think):

https://github.com/flutter/flutter/issues/69429#issuecomment-906627145

Essentially, loading each test file is taking longer for me than running the tests in that file.

16 tests in one file: `00:05 +16: All tests passed!`
16 tests split between 4 files: `00:08 +16: All tests passed!`
16 tests with one test per file: `00:22 +16: All tests passed!`

**Comment by AlexMcConnell** [comment]
The problem has actually gotten worse. The above stats were done in Flutter 2.0.6, though upgrading to Flutter 2.2.3 provided the same stats. But...

Flutter dev - 2.6.0-0.0.pre
16 tests in one file: `00:05 +16: All tests passed!`
16 tests split between 4 files: `00:09 +16: All tests passed!`
16 tests with one test per file: `00:26 +16: All tests passed!`

Flutter master - 2.6.0-1.0.pre.105
16 tests in one file: `00:05 +16: All tests passed!`
16 tests split between 4 files: `00:09 +16: All tests passed!`
16 tests with one test per file: `00:26 +16: All tests passed!`

**Comment by jakemac53** [comment]
Are these flutter tests or regular dart vm tests you are running?

**Comment by AlexMcConnell** [comment]
Flutter widget tests, which I expect have their own special overhead. But the imports aren't drastically different.

**Comment by svarlet** [comment]
I shared some numbers in https://github.com/flutter/flutter/issues/69429#issuecomment-903582893. Though I didn't specify that we have 62 test files (359 tests in total, 14 seconds total). These are vm, non widget, tests. A very limited few of them have file system side effects but they run in an acceptable amount of time (coming from someone who'd like a total timing < 1s).

For people who run their tests very often (every other minute for example), the amount spent waiting is unbearably long. Consequences are loss of focus, running tests less often (if ever), difficulty to teach the benefits of automated testing to juniors, devs pushing broken code, build breaking, developers stuck after pulling broken code, etc...). 

For example, you said that `pkg/test_api` runs 17 files / 305 tests in 9 seconds, (which is surprisingly similar to our scale/timings). It's at least 5 to 10 times longer that it should to avoid the negative consequences enumerated above. Many languages and their test framework offer an order of magnitude timings, without even involving concurrency (and this test pkg has concurrency built in!). Are there design flaws in this test pkg that limits its performance? Is there another test framework for dart programmers without these flaws?

**Comment by jakemac53** [comment]
The flutter widget tests use their own platform for loading, so for those you would want to keep trying to push on the flutter issue. You can find that code I believe here, https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/test/flutter_platform.dart.

In terms of general test running times, Dart is a compiled language and most of the time here is just compiling the tests. There is a limit to how fast we can make things in the current model, although there are some potential improvements that could still be made. But <1s to run all your tests is likely not achievable - that is about how much time we spend doing an incremental recompile of package:test itself today for instance, so even the most basic of tests would take that long just to rebuild.

There are definitely improvements that could be made over the current system though. Today we don't store any state between builds (so each run does a full recompile). And you can trigger some pathological behavior if your tests have very different transitive imports (you can end up loading, invalidating, and reloading the same sources).

**Comment by jakemac53** [comment]
I filed https://github.com/dart-lang/test/issues/1570 to track investigation into lightweight isolates. But I wouldn't expect much traction on that in the short term, as it would be a large body of work.

**Comment by AlexMcConnell** [comment]
Why would isolates even be involved?

**Comment by jakemac53** [comment]
The main test runner is the thing that drives running all the tests. This is what enables a lot of the features of the test runner, in particular the multi platform support but also all test suite and global config is handled through that.

It needs to then find your tests and run them on the desired platform, and for running Dart VM tests using an Isolate makes a lot of sense.

**Comment by AlexMcConnell** [comment]
But is it using an isolate for each test file? That might explain the speed difference between normal numbers of tests per file vs putting them all in one file.

**Comment by jakemac53** [comment]
> But is it using an isolate for each test file?

Yes it does - I outlined this as one of the problems in the issue I filed. In order to use lightweight isolates we would have to merge them all into one file anyways.

Note though that we would still have to run each test in its own isolate, because each one needs to have a clean global state. If one test changed some global state that shouldn't leak into the others. But the overhead of launching isolates would be dramatically reduced with the light weight isolates.
